Collect data with Elastic Cloud Native OpenTelemetry Endpoint (#4854)
elastic/observability-docs: docs/en/serverless/quickstarts/collect-data-with-native-otlp.asciidoc

[discrete]
== Get started

[discrete]
=== Get your native OTLP endpoint credentials

. <<observability-create-an-observability-project,Create a new {obs-serverless} project>>, or open an existing one.

. In your {obs-serverless} project, go to **Add Data**.

. Under **What do you want to monitor?** select **Application**, and then under **Monitor your application using** select **OpenTelemetry**.
+
NOTE: Follow this flow for all use cases, including logs and infrastructure monitoring.

. Copy the `OTEL_EXPORTER_OTLP_ENDPOINT` URL. Replace `.apm` with `.ingest` and save this value for later.

[discrete]
=== Create an API key

. Click "Create an API Key" to generate a new API key. Copy this value for later.
. (Optional) Test your new API key. You can do this by sending an empty JSON object to the `/v1/traces` endpoint. For example:
+
[source,bash]
----
curl -X POST \
  -H "Content-Type: application/json" \
  -H "Authorization: ApiKey <api-key>" \
  https://{YOUR_CLUSTER}.ingest.us-east-1.aws.elastic.cloud:443/v1/traces \
  -d '{}'
----
+
The response should be similar to:
+
[source,txt]
----

> {"partialSuccess":{}}%  

Author: This is the result I got when testing this. Does this sound right?

Reviewer: Looks good.

----

[discrete]
=== Send data to your native OTLP endpoint

> 

Author: If you have an OTEL collector up and running, configure the exporter with this data

Author: Clarify `"Authorization: ApiKey <api-key>"`

To send data to your native OTLP endpoint, add the above credentials to your OpenTelemetry collector or SDK. If you're unsure how to accomplish this, the following links may be helpful: 

* https://opentelemetry.io/docs/collector/configuration/[OTLP Collector configuration]
* https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/[OTLP Exporter configuration]
* https://elastic.github.io/opentelemetry/edot-collector/edot-collector-config.html[Elastic Distributions of OpenTelemetry (EDOT) Collector configuration]
* https://elastic.github.io/opentelemetry/edot-sdks/index.html[Elastic Distributions of OpenTelemetry (EDOT) SDKs]

// I'm not entirely sure how to be more helpful here
// Should we expand on some of these in this doc?

That's it. You're now ready to leverage OTel in Elastic Observability.

[discrete]

> ==== (Optional) Install EDOT Collector

Author: This is #2

Don't have a collector or SDK running? No problem. Spin up an EDOT collector in just a few steps:

* https://elastic.github.io/opentelemetry/quickstart/serverless/k8s.html[Kubernetes Quickstart]
* https://elastic.github.io/opentelemetry/quickstart/serverless/hosts_vms.html[Hosts & VMs Quickstart]

> // Commenting out Docker until the docs are ready

Author: APM coming soon

// * https://elastic.github.io/opentelemetry/quickstart/serverless/docker.html[Docker Quickstart]

[discrete]

> == Get creative with Discover

Author: Move this to a new page in apps and services

Make for ANY otlp data

Author: End goal is analysis of OTLP data in one place

The Applications UI allows you to monitor your software services and applications in real-time. You can visualize detailed performance information on your services, identify and analyze errors, and monitor host-level metrics.

[discrete]
== Check the health of your infrastructure

To access the **Hosts** page, in your {obs-serverless} project, go to
**Infrastructure** → **Hosts**.

On the Hosts page, you can view health and performance metrics to help you quickly:

* Analyze and compare hosts without having to build new dashboards.
* Identify which hosts trigger the most alerts.
* Troubleshoot and resolve issues quickly.

[role="screenshot"]
image::images/hosts-ui-otlp.png[resource attributes]

See {observability-guide}/observability-analyze-hosts.html[Analyze and compare hosts] to learn more.

[discrete]
=== (Optional) Install the OpenTelemetry Assets integration

Install the OpenTelemetry Assets integrations to access the "[OTEL][Metrics Kubernetes] Cluster Overview" dashboard.
This dashboard 

In your {obs-serverless} project, go to **Integrations** and toggle **Display beta integrations**.
Search for "OpenTelemetry" and select and install **Kubernetes OpenTelemetry Assets**.

> // Does anyone have a cluster with relevant data that we can use to take a screenshot of this dashboard?

Author: Asked above too

> Does anyone have a cluster with relevant data that we can use to take a screenshot of this dashboard?

Reviewer: @mlunadia ^^ could you provide it since you have already have one running. Thanks

[discrete]
== Troubleshoot

> // are there specific troubleshooting scenarios we should call out?

Author: > are there specific troubleshooting scenarios we should call out?

Author: "Authorization: ApiKey <api-key>"

Author: @vigneshshanmugam any thoughts on this topic?

Reviewer: Nothing specific comes to mind as you already called out the ApiKey format. I am also thinking if we can move the ratelimit scenario in troubleshooting as opposed to limitation. 

```
Error: too many requests

The Managed endpoint has per-project rate limits in place. If you hit this limit, reach out to our https://support.elastic.co[support team].
```

Maybe point the user to limitation doc if they are seeing any other errors? We can add other guides as followups as we learn more.

[[collect-data-with-native-otlp]]

> = Quickstart: Collect data with Elastic Cloud Native OpenTelemetry Endpoint

Reviewer 2: ```suggestion
= Quickstart: Send data to the Elastic Cloud Managed OTLP Endpoint
```

Author: I'll make this change but are we okay with assuming that folks know OTLP == OpenTelemetry?

preview::[]

> In this quickstart guide, you'll learn how to use the Elastic Cloud Native OpenTelemetry Endpoint to send logs, metrics, and traces to Elastic.

Reviewer 2: ```suggestion
In this quickstart guide, you'll learn how to use the Elastic Cloud Managed OTLP Endpoint to send logs, metrics, and traces to Elastic.
```

[discrete]
== What is the Elastic Cloud Native OpenTelemetry Endpoint?

> The Elastic Cloud Native OpenTelemetry Endpoint is a fully managed offering exclusively for Elastic Cloud Serverless users that simplifies OpenTelemetry data ingestion. It provides an endpoint for OpenTelemetry SDKs and Collectors to send telemetry data, with Elastic handling scaling, data processing, and storage.

Reviewer 2: ```suggestion
The Managed OTLP Endpoint is a fully managed offering exclusively for Elastic Cloud users (initially available in Elastic Cloud Serverless only) that simplifies OpenTelemetry data ingestion. It provides an endpoint for OpenTelemetry SDKs and Collectors to send telemetry data, with Elastic handling scaling, data processing, and storage.
```

This endpoint is designed for the following use cases:

* Logs & Infrastructure Monitoring: Logs forwarded in OTLP format and host and Kubernetes metrics in OTLP format.
* APM: Application telemetry in OTLP format.

[%collapsible]

> .Differences from the existing Elastic OTLP Endpoint

Reviewer 2: ```suggestion
.Differences from the existing Elastic APM Endpoint
```

====

> The Elastic Cloud Native OpenTelemetry Endpoint ensures that OpenTelemetry data is stored without any schema translation, preserving both semantic conventions and resource attributes. It supports ingesting OTLP logs, metrics, and traces in a unified manner, ensuring consistent treatment across all telemetry data. This marks a significant improvement over the {observability-guide}/apm-open-telemetry.html[existing functionality], which primarily focuses on traces and the APM use case.

Reviewer 2: ```suggestion
The Elastic Cloud Managed OTLP Endpoint ensures that OpenTelemetry data is stored without any schema translation, preserving both semantic conventions and resource attributes. It supports ingesting OTLP logs, metrics, and traces in a unified manner, ensuring consistent treatment across all telemetry data. This marks a significant improvement over the {observability-guide}/apm-open-telemetry.html[existing functionality], which primarily focuses on traces and the APM use case.
```

> The Elastic Cloud Native OpenTelemetry Endpoint ensures that OpenTelemetry data is stored without any schema translation, preserving both semantic conventions and resource attributes. It supports ingesting OTLP logs, metrics, and traces in a unified manner, ensuring consistent treatment across all telemetry data. This marks a significant improvement over the {observability-guide}/apm-open-telemetry.html[existing functionality], which primarily focuses on traces and the APM use case.

Reviewer: resource attributes is part of semantic convention. Maybe reword it as `Opentelemetry semantic conventions`?

Author: updated

====

[discrete]
== Prerequisites

* An {obs-serverless} project. To learn more, refer to <<observability-create-an-observability-project>>.
* A system forwarding logs, metrics, or traces in OTLP (any OTel Collector or SDK—EDOT or community).

[discrete]
=== Limitations

> * The Elasticsearch exporter only supports histograms with delta temporality. Cumulative histograms are dropped.

Reviewer 2: This a limitation but it needs to be expressed in the context of the endpoint, whilst we know otel collector/elasticsearch exporter is behind it, users don't and they don't need to.

> * You may be rate limited if...

Reviewer: We have per-project rate limits, maybe go via the support/SDH route to increase it? @mlunadia thoughts?

Author: Updated

> * Resolution concerns may arise for explicit bounded histogram metrics due to Memory issues.

Reviewer: Latency distributions based on histogram values have limited precision due to the fixed boundaries of explicit bucket histograms.

Author: Have you considered a career switch to technical writing?

[discrete]
== Get started

[discrete]

> === Get your native OTLP endpoint

Reviewer 2: ```suggestion
=== Get your native OTLP endpoint credentials
```

. <<observability-create-an-observability-project,Create a new {obs-serverless} project>>, or open an existing one.

. In your {obs-serverless} project, go to **Add Data**.

. Under **What do you want to monitor?** select **Application**, and then under **Monitor your application using** select **OpenTelemetry**.
+
NOTE: Follow this flow for all use cases, including logs and infrastructure monitoring.

> . Copy the `OTEL_EXPORTER_OTLP_ENDPOINT` URL. Replace `.apm` with `.ingest` and save this value for later.

Reviewer 2: Users will no longer have to edit the endpoint, the actual endpoint url will be displayed in the [cloud console](https://github.com/elastic/cloud/issues/137354) and also in the application onboarding. I recommend we guide users to obtain the endpoint from cloud console.

Reviewer 2: ![image](https://github.com/user-attachments/assets/9ba83011-0621-4796-ba6d-ca1f552e4232)

Author: To clarify: We should recommend users go to the cloud console for both the Endpoint and API key, not the onboarding flow?

Should we doc both?

Author: Updated to:  "Managed OTLP endpoint"

Author: @mlunadia I'm not able to replicate this page in the cloud console.

When I click on "Manage project API keys" I'm taken to this page 
<img width="650" alt="Screenshot 2025-03-26 at 8 03 06 PM" src="https://github.com/user-attachments/assets/1306d955-dca5-4cf6-a57d-5691c9a1528d" />

[discrete]
=== Create an API key

> . Click "Create an API Key" to generate a new API key. Copy this value for later.

Reviewer 2: If go with the cloud console route this means users can generate the api key there just with an extra step.

Reviewer 2: ![image](https://github.com/user-attachments/assets/89f6eb93-17f7-459a-9473-7362b7e6fe2a)

. (Optional) Test your new API key. You can do this by sending an empty JSON object to the `/v1/traces` endpoint. For example:
+
[source,bash]
----
curl -X POST \
  -H "Content-Type: application/json" \
  -H "Authorization: ApiKey <api-key>" \
  https://{YOUR_CLUSTER}.ingest.us-east-1.aws.elastic.cloud:443/v1/traces \
  -d '{}'
----
+
The response should be similar to:
+
[source,txt]
----
{"partialSuccess":{}}%  
----

[discrete]
=== Send data to your native OTLP endpoint

To send data to your native OTLP endpoint, add the above credentials to your OpenTelemetry collector or SDK. If you're unsure how to accomplish this, the following links may be helpful: 

* https://opentelemetry.io/docs/collector/configuration/[OTLP Collector configuration]
* https://opentelemetry.io/docs/languages/sdk-configuration/otlp-exporter/[OTLP Exporter configuration]
* https://elastic.github.io/opentelemetry/edot-collector/edot-collector-config.html[Elastic Distributions of OpenTelemetry (EDOT) Collector configuration]
* https://elastic.github.io/opentelemetry/edot-sdks/index.html[Elastic Distributions of OpenTelemetry (EDOT) SDKs]

> // I'm not entirely sure how to be more helpful here

Reviewer 2: seems good enough for me

// Should we expand on some of these in this doc?

That's it. You're now ready to leverage OTel in Elastic Observability.

[discrete]
==== (Optional) Install EDOT Collector

> Don't have a collector or SDK running? No problem. Spin up an EDOT collector in just a few steps:

Reviewer 2: We can send them to the add data flow or [edot quickstart ](https://elastic.github.io/opentelemetry/quickstart/serverless/)

Author: Awesome. I didn't know this existed. Same comment as below:

> When we move the EDOT docs to our new docs system we'll easily be able to single-source this content into both locations. For now, I'll link folks to the EDOT page.

      Kubernetes (Operator)
    </button>
    <button role="tab"
            aria-selected="false"
            aria-controls="linux-tab"
            id="linux-tab-id"
            tabindex="-1">
      Linux (EDOT Collector)
    </button>
    <button role="tab"
            aria-selected="false"
            aria-controls="mac-tab"
            id="mac-tab-id"
            tabindex="-1">
      Mac (EDOT Collector)
    </button>
  </div>
  <div tabindex="0"
       role="tabpanel"
       id="helm-tab"
       aria-labelledby="helm-tab-id">
++++

include::native-otlp-tabs.asciidoc[tag=helm]

++++
  </div>
  <div tabindex="0"
       role="tabpanel"
       id="linux-tab"
       aria-labelledby="linux-tab-id"
       hidden="">
++++

include::native-otlp-tabs.asciidoc[tag=linux]

++++
  </div>
  <div tabindex="0"
       role="tabpanel"
       id="mac-tab"
       aria-labelledby="mac-tab-id"
       hidden="">
++++

include::native-otlp-tabs.asciidoc[tag=mac]

++++
  </div>
</div>
++++

[discrete]
== Get Creative with Discover

Attributes and resource attributes are prefixed with `attributes.*` and `resource.attributes.*`.
You can correlate all signals with a single `resource.attributes.*`.

> // screenshots here?

Author: Addded

// is there a specific workflow to call out?

[discrete]
== Monitor Application Performance

// I'm guessing these are the UIs that work with these attrs?

* UIs
* Service Inventory
* Service Map
* Service Groups
* Service Specific Views
** Overview
** Transactions
** Dependencies
** Errors
** Metrics
** Logs
** Infrastructure
** Traces
** Dependencies

> // screenshots here?

Author: ++ need some input on what to do with this section. Added a brief overview on our APM UI for now.

// is there a specific workflow to call out?

[discrete]
== Check the Health of Your Infrastructure

* Hosts
** Hosts UI
* Kubernetes
** (Optional) Install the OpenTelemetry Assets integrations to access the ‘[OTEL][Metrics Kubernetes] Cluster Overview’ dashboard.

> // screenshots here?

Author: Is there a cluster with relevant data that I can use to take this screenshot?
